controllers/components/kserve/kserve_support.go

```diff
@@ -0,0 +1,300 @@
+	"fmt"
+	"path"
+
+	"github.com/hashicorp/go-multierror"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	componentsv1 "github.com/opendatahub-io/opendatahub-operator/v2/apis/components/v1"
+	dsciv1 "github.com/opendatahub-io/opendatahub-operator/v2/apis/dscinitialization/v1"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/cluster"
+	odhtypes "github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/types"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/deploy"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/feature"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/feature/manifest"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/feature/serverless"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/feature/servicemesh"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/resources"
+)
+
+var serviceAccounts = map[cluster.Platform][]string{
+	cluster.Unknown:          {odhModelControllerComponentName},
+	cluster.OpenDataHub:      {odhModelControllerComponentName},
+	cluster.ManagedRhods:     {odhModelControllerComponentName},
+	cluster.SelfManagedRhods: {odhModelControllerComponentName},
+}
+
+var clusterRolePredicate = predicate.Funcs{
+	UpdateFunc: func(e event.UpdateEvent) bool {
+		notAllowedNames := []string{"kserve-admin"}
+		for _, notallowedName := range notAllowedNames {
+			if e.ObjectNew.GetName() == notallowedName {
+				return false
+			}
+		}
+		return true
+	},
+}
+
+var compareHashPredicate = predicate.Funcs{
+	UpdateFunc: func(e event.UpdateEvent) bool {
+		oldUnstructured, err := resources.ToUnstructured(e.ObjectOld.DeepCopyObject())
+		if err != nil {
+			return true
+		}
+		newUnstructured, err := resources.ToUnstructured(e.ObjectNew.DeepCopyObject())
+		if err != nil {
+			return true
+		}
+
+		oldHash, err := resources.Hash(oldUnstructured)
+		if err != nil {
+			return true
+		}
+		newHash, err := resources.Hash(newUnstructured)
+		if err != nil {
+			return true
+		}
+
+		return base64.RawURLEncoding.EncodeToString(oldHash) != base64.RawURLEncoding.EncodeToString(newHash)
+	},
+}
+
+func kserveManifestInfo(sourcePath string) odhtypes.ManifestInfo {
+	return odhtypes.ManifestInfo{
+		Path:       deploy.DefaultManifestPath,
+		ContextDir: componentName,
+		SourcePath: sourcePath,
+	}
+}
+
+func odhModelControllerManifestInfo(sourcePath string) odhtypes.ManifestInfo {
+	return odhtypes.ManifestInfo{
+		Path:       deploy.DefaultManifestPath,
+		ContextDir: odhModelControllerComponentName,
+		SourcePath: sourcePath,
+	}
+}
+
+func checkDependentOperators(ctx context.Context, cli client.Client) *multierror.Error {
```

Comment:
maybe this should be made a pre-requisite check ? like https://github.com/opendatahub-io/opendatahub-operator/blob/acd89b8846393d17e0715cafce00237961911261/pkg/controller/predicates/dependent/dependent.go#L78-L114

Reply:
I've pulled this out into an initial action to check that will block if those operators aren't installed in the case where they're needed. I'll mark this comment thread as resolved, and any further discussion can happen in a new comment thread on the updated code.